SMTP Network Filter in contrib extensions

VishalDamgude opened the pull request:
Commit Message: Smtp Network Filter in contrib
Additional Description:

This smtp network filter has been developed as per RFC 5321 (https://www.rfc-editor.org/rfc/rfc5321) and supports STARTTLS extension. It will parse SMTP commands and responses as per the SMTP protocol and provide protocol specific metrics.
The downstream starttls is supported by default. When smtp client sends STARTTLS, the filter will start secure transport on downstream connection. The upstream encryption is configurable. If enabled and if TLS handshake fails during upstream TLS negotiation, the filter responds to the client with error and closes the connection. If upstream TLS handshake succeeds, both the downstream and upstream connections are encrypted for secure smtp mail transfer.

Note: This new PR is created due to CI kickstart issues with another PR: https://github.com/envoyproxy/envoy/pull/25398

Risk Level: Low
Testing: Unit Tests, Manual Testing
Docs Changes:
Release Notes:
Platform Specific Features: No
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/24219
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

wbpcode commented on contrib/smtp_proxy/filters/network/source/config.cc:
```diff
@@ -0,0 +1,35 @@
+NetworkFilters::SmtpProxy::SmtpConfigFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::network::smtp_proxy::v3alpha::SmtpProxy& proto_config,
+    Server::Configuration::FactoryContext& context) {
+  ASSERT(!proto_config.stat_prefix().empty());
```
I think this ASSERT is unnecessary? This condition should be ensured by the proto validation.

VishalDamgude replied:
Removed it.

wbpcode commented on contrib/smtp_proxy/filters/network/source/smtp_decoder.cc:
```diff
@@ -0,0 +1,299 @@
+namespace NetworkFilters {
+namespace SmtpProxy {
+
+Decoder::Result DecoderImpl::onData(Buffer::Instance& data, bool upstream) {
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+
+  if (upstream) {
+    result = parseResponse(data);
+    data.drain(data.length());
+    return result;
+  }
+  result = parseCommand(data);
+  data.drain(data.length());
+  return result;
+}
+
+Decoder::Result DecoderImpl::parseCommand(Buffer::Instance& data) {
+  ENVOY_LOG(debug, "smtp_proxy: decoding {} bytes", data.length());
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+  std::string command = data.toString();
+
+  if (command.length() < 6) {
+    // Message size is not sufficient to parse.
+    return result;
+  }
```
If the command length is less then 6, then the parseCommand will do nothing return deirectly. And then the `onData` call will drain the unparsed buffer? I think here are some error, or I miss something?

VishalDamgude replied:
If command length is <6, currently the smtp filter will pass it to next filter (tcp filter) and it will reach upstream. Upstream can reject the command with 4xx bad syntax. But we can have smtp filter itself reject commands with length < 6. Will do this change

cpakulski replied:
> If command length is <6, currently the smtp filter will pass it to next filter (tcp filter) and it will reach upstream. Upstream can reject the command with 4xx bad syntax. But we can have smtp filter itself reject commands with length < 6. Will do this change

Do not do it. Upstream server has its own buffering. The logic behind buffering partial packets is to address the fact that packets may arrive to the filter in unknown chunks. For example, it filter receives 3 bytes, pass those 3 bytes upstream and continue buffering until filter has a protocol unit which can be parsed. In the meantime, keep passing upstream whatever arrives from downstream.
In postgres I used logic to wait for minimum number of bytes, because that represented postgres header and each packet has that header. Not sure about smtp.

VishalDamgude replied:
Hey @cpakulski ,
I refactored the logic to parse smtp commands and responses. Please have a look at DecoderImpl::parseCommand() function.

wbpcode replied:
I didn't think @cpakulski 's (and my) concern is be addressed. From the source code, the input buffer will always be drained even if there is no complete command and nothing is processed accutally.
Or I missed something?

/wait

VishalDamgude replied:
@wbpcode Got you point. I am doing data.drain() each time e.g.
  `result = parseCommand(data);
  data.drain(data.length());`
  Will make changes to buffer the received bytes if they are not sufficient to parse.

wbpcode reviewed (commented):
Thanks for you contribution. Some comments are added to start this review.

cpakulski reviewed (changes requested):
Please add unit tests to check how messages are buffered until something which can be parsed is in the buffer.

jsbucy commented:
I'm happy that Vishal is taking the lead here but I would like to see a little more discussion about the requirements perhaps on #19765

jmarantz commented:
@adisuissa ping

adisuissa commented on api/contrib/envoy/extensions/filters/network/smtp_proxy/v3alpha/smtp_proxy.proto:
```diff
@@ -0,0 +1,47 @@
+  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
+
+  // If enabled, filter will generate x-req-id to identify smtp session/transaction and send it to upstream.
+  bool tracing = 2;
```
Taking a step back, should the [UUID request_id](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/request_id/uuid/v3/uuid.proto.html#uuid-proto) be configured to achieve tracing?

adisuissa commented on api/contrib/envoy/extensions/filters/network/smtp_proxy/v3alpha/smtp_proxy.proto:
```diff
@@ -0,0 +1,47 @@
+// <config_network_filters_smtp_proxy>`.
+// [#extension: envoy.filters.network.smtp_proxy]
+
+message SmtpProxy {
```
Does this require the use of a TcpProxy filter?

VishalDamgude replied:
Yes, We require a TCPProxy as terminal filter.

adisuissa reviewed (commented):
Thanks.
Left a few high-level comments/questions about the API.

Note that this PR is pretty large, and although contrib code has somewhat of a lower bar, it makes reviewing the code changes challenging. Please consider dissecting it to smaller PRs.

alyssawilk commented:
/wait on reviewer comments

VishalDamgude commented on api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:
```diff
@@ -611,8 +611,7 @@ message FilterStateRule {
   // in the FilterState with the name specified in the ``name`` field above.
-  map<string, JwtRequirement>
-  requires = 3;
+  map<string, JwtRequirement> requires = 3;
```
Thanks for the comments @adisuissa , Will address and update the PR.

wbpcode commented:
@VishalDamgude any relationship between this PR and #27437? Seems there are some repeated content?
cc @jsbucy

adisuissa commented:
/wait-any

VishalDamgude commented:
> @VishalDamgude any relationship between this PR and #27437? Seems there are some repeated content? cc @jsbucy

@wbpcode Yes, the content seems repeated in #27437. The STARTTLS functionality (both upstream and downstream TLS support) is already implemented as part of this PR along with additional features like metrics and tracing.

@jsbucy if any additional functionality is required, can we please add it to this PR itself.

I will be addressing the review comments in this PR this week.

jsbucy commented:
Hi Vishal, glad to connect! I wrote a [doc ](https://docs.google.com/document/d/1F2yRVulML-qdoQkXzWIv5DPumKDSTvIQGr-MZxluEQ0/edit) detailing my assumptions for STARTTLS. I think the implementation in #27437 supports some more use cases that I think are important, in particular to sidecar with an smtp client or server (forward or reverse proxy) that doesn’t implement the STARTTLS extension.

One possible way forward would be to decouple the STARTTLS/session startup part from the deep protocol inspection/tracing part. Does the implementation here support the SMTP CHUNKING extension rfc3030?

wbpcode commented:
Because these are lots of repeated content, could you figure out which one should be landed first? This is a huge work and no maintainer is familiar to the SMTP protocol, so, it's pretty hard for comunity to review both PRs or do a decision.
cc @VishalDamgude @jsbucy

wbpcode commented:
/wait

jsbucy commented:
Vishal and I had a call last night to discuss, here are some general notes from the call, I'll add some more specific discussion topics in separate comments.

Vishal’s org is using this for reverse proxy with a custom SMTP server that doesn’t advertise many smtp extensions
tracing/audit loging is a requirement for them

I’m fine with landing this PR first and iterating
but I would like to try to avoid implementing backend protocol features that are completly unique to their application, (further comments to follow about x-req-id, starttls)
I can port over the deltas from my PR once this has landed, in particular the support for additional starttls configs (upstream initiation for forward proxy, downstream termination transparent to the upstream) and the integration test

jsbucy commented:
I think the current state of the tracing/deep protocol inspection code has some limitations and it will take some iteration to make it work with an MTA such as postfix that implements a typical suite of smtp extensions. I think it needs to be possible to disable protocol inspection after STARTTLS (corresponding to the PASSTHROUGH state in my PR) and that should be the default
I think it’s fair to say firewalls have had a lot of trouble getting this right in the past
one way this can go wrong is the inspector gets out of sync with the session and ends up interpreting an email message as smtp commands and hilarity ensues
possibly the inspection code should have some heuristics to disable itself on too many unknown commands or if the server/upstream advertises an unknown/unsupported extension (I think PIPELINING, CHUNKING, AUTH PLAIN all definitely won’t work now)
we should make sure something like AUTH PLAIN won’t end up logging the password as an unknown command in the envoy audit log!

jsbucy commented:
would the postfix [XCLIENT](https://www.postfix.org/XCLIENT_README.html) extension work for you instead of x-req-id? 
I was planning on adding this once we get basic starttls done

jsbucy commented:
STARTTLS, in the current state of the code here:
- the server must advertise the capability
- the client must initiate STARTTLS command
- if client/downstream sends starttls and server/upstream returns a 2xx response, the filter enables downstream tls
if upstream tls is configured, also does enables it there
if upstream tls is disabled, this is basically a custom protocol where the server advertises the extension and returns a 2xx response but doesn’t actually do TLS?
are you using it this way? does your application depend on the server receiving the starttls command? would passing the TLS info in XCLIENT work for you?
do the possible starttls modes implemented in my PR meet your needs here?

wbpcode commented:
cc @jsbucy Thanks very much for your hard work and response.

Then could you help to review this PR also? I think we still need some owner to maintain this extension after it's land.

jsbucy commented:
> cc @jsbucy Thanks very much for your hard work and response.
> 
> Then could you help to review this PR also? I think we still need some owner to maintain this extension after it's land.

Yeah I'll take another look at this PR.

I was expecting to be the owner if you don't mean an official envoy maintainer. The thing I really want to build is https://github.com/envoyproxy/envoy/issues/9133 for which starttls is basically the first step. I've been doing some prototyping to to figure out what that looks like before spending a lot of time building the production-grade envoy filter.

VishalDamgude commented:
Attaching a document outlining all the features included in this PR and the proposed changes:
https://docs.google.com/document/d/151tzHSvzs-xSyTJ7HvcepJsPTvO8mxVMNvnpAKDgmgg/edit?usp=sharing

jsbucy commented:
@VishalDamgude are you using this with upstream_ssl enabled or disabled?

jsbucy commented on contrib/smtp_proxy/filters/network/source/smtp_decoder_impl.cc:
```diff
@@ -0,0 +1,112 @@
+
+  std::string command = "";
+  std::string args = "";
+  if (length < 4) {
```
I don't think this is going to be able to transparently handle unknown commands since it could be something like BDAT or AUTH/SASL modes (other than PLAIN) that take the protocol out of "line mode".

I would just compare to a list of the commands you know how to handle and disable protocol inspection if you see anything else.

jsbucy commented on contrib/smtp_proxy/filters/network/source/smtp_decoder_impl.cc:
```diff
@@ -0,0 +1,112 @@
+  ENVOY_LOG(debug, "received command {}", buffer);
+  // Each SMTP command ends with CRLF ("\r\n"), if received buffer doesn't end with CRLF, the filter
+  // will not process it.
+  if (!absl::EndsWith(buffer, SmtpUtils::smtpCrlfSuffix)) {
```
does this work if you get the command split across multiple reads? corresponding to NeedMoreData in postgres Decoder::Result

jsbucy commented on contrib/smtp_proxy/filters/network/source/smtp_decoder_impl.cc:
```diff
@@ -0,0 +1,112 @@
+    return result;
+  }
+
+  buffer = StringUtil::cropRight(buffer, SmtpUtils::smtpCrlfSuffix);
```
I would find the first CRLF and if there is anything after it, the client is doing pipelining or sending junk so disable protocol inspection.

VishalDamgude replied:
Sure, modifying the parsing logic.

VishalDamgude commented:
WIP.

cpakulski commented:
There are 2 open PRs for SMTP. Is there any concensus which one should go ahead? See https://github.com/envoyproxy/envoy/pull/27437.

cpakulski commented:
ping. I guess we want to keep it open.

VishalDamgude commented:
ping. please remove stale label

VishalDamgude commented:
/wait-any

VishalDamgude commented:
/wait-any

adisuissa commented:
/wait

VishalDamgude commented:
/wait-any

VishalDamgude commented:
/wait-any
